luatest/server_v2.lua

```diff
@@ -0,0 +1,493 @@
+            end
+            if clock.time() > deadline then
+                error(('Stopping of server (alias: %s, workdir: %s, pid: %d) timed out')
+                    :format(self.alias, fio.basename(self.workdir), pid))
+            end
+            fiber.sleep(WAIT_DELAY)
+        end
+    end
+end
+
+--- Cleanup the server working directory.
+function Server:cleanup()
+    fio.rmtree(self.workdir)
+    self.instance_id = nil
+    self.instance_uuid = nil
+end
+
+--- Stop the server and cleanup its working directory.
+function Server:drop()
+    self:stop()
+    self:cleanup()
+end
+
+-- Wait until the given condition is `true` (anything except `false` and `nil`).
+-- Throw an error when timeout exceeds.
+local function wait_for_condition(cond_name, server, func, ...)
+    local pid = server.process.pid
+    local deadline = clock.time() + WAIT_TIMEOUT
+    while true do
+        if func(...) then
+            return
+        end
+        if clock.time() > deadline then
+            error(('Waiting for "%s" condition on server (alias: %s, workdir: %s, pid: %d) timed out')
+                :format(cond_name, server.alias, fio.basename(server.workdir), pid))
+        end
+        fiber.sleep(WAIT_DELAY)
+    end
+end
+
+--- Wait until the server is ready after the start.
+-- It is considered that the server is operable when `_G.ready` variable becomes `true` on it.
+function Server:wait_until_ready()
+    return wait_for_condition('server is ready', self, function()
+        local ok, is_ready = pcall(function()
+            self:connect_net_box()
+            return self.net_box:eval('return _G.ready') == true
+        end)
+        return ok and is_ready
+    end)
+end
+
+--- Get ID of the server instance.
+--
+-- @return number
+function Server:get_instance_id()
+    -- Cache the value when found it first time.
+    if self.instance_id then
+        return self.instance_id
+    end
+    local id = self:exec(function() return box.info.id end)
+    -- But do not cache 0 - it is an anon instance, its ID might change.
+    if id ~= 0 then
+        self.instance_id = id
+    end
+    return id
+end
+
+--- Get UUID of the server instance.
+--
+-- @return string
+function Server:get_instance_uuid()
+    -- Cache the value when found it first time.
+    if self.instance_uuid then
+        return self.instance_uuid
+    end
+    self.instance_uuid = self:exec(function() return box.info.uuid end)
+    return self.instance_uuid
+end
+
+--
+-- Log
+--
+
+--- Grep the server logfile for a string pattern.
+-- If the server has crashed, it is required to provide `opts.filename`.
+--
+-- @string pattern String pattern to search in the server logfile.
+-- @number[opt] bytes_num Number of bytes to read from the server logfile.
+-- @tab[opt] options
+-- @bool[opt] options.reset Reset search if the server was restarted. Defaults to `true`.
+-- @string[opt] options.filename Path to the server logfile. Defaults to `box.cfg.log`.
+-- @return string|nil
```

Comment:
> search if the server was restarted

does it mean 'search even if it was restarted'? Or 'also search in logs before the restart'?


```suggestion
--- Grep the server's log file with a string pattern.
-- If the server has crashed, the `opts.filename` is requred.
--
-- @string pattern String pattern to search for in the server's log file.
-- @number[opt] bytes_num Number of bytes to read from the server's log file.
-- @tab[opt] options
-- @bool[opt] options.reset Reset search if the server was restarted. Defaults to `true`.
-- @string[opt] options.filename Path to the server log file. Defaults to `box.cfg.log`.
-- @return string|nil
```

Reply:
> Also, opts.filename is required, but it also has a default value. Should the developer override that value when a server has crashed?

Yes, because there is no server to execute `box.cfg.log`.

Reply:
Rephrased a little. Fixed.

Reply:
Nice!